teps/0021-results-api.md

```diff
@@ -0,0 +1,772 @@
+existing TaskRun/PipelineRun API backed by durable storage. While this would use
+the existing Tekton Pipelines types and make it easier to have existing clients
+(dashboard, tkn) migrate to the new service, this would not cover the everything
+that happens as part of an event:
+
+- User configured DSLs may not fit neatly into the TaskRun / PipelineRun format,
+  but they are useful to retain to describe what the user requested.
+- Incoming Trigger events do not have a home in Task/PipelineRuns today. We
+  likely do not want to add this due to
+  [restrictions on Kubernetes object sizes](https://stackoverflow.com/a/53015758) -
+  as an example,
+  [GitHub webhook events can be up to 25 MB](https://developer.github.com/webhooks/event-payloads/#webhook-payload-object-common-properties).
+- This does not allow for post-run operation statuses like Cloud Event
+  publishing, Pull Request updates, etc. to be associated to their execution
+  events. It's unclear whether we would want to store this information in Runs,
+  since it is a different resource / execution.
+
+We may choose to provide a facade of the Task/Pipeline APIs in the future for
+convenience, but this would be a layer in addition to the Results API.
+
+### REST/Open API
+
+One alternative to using gRPC would be to define an Open API spec and implement
+a REST API as the surface.
+
+We are choosing to go with gRPC to take advantage of protobuf serialization,
+multi-language client/server library generation, and other performance features
+baked in to gRPC by default.
+
+For clients that require a REST API, we can configure a
+[gateway](https://github.com/grpc-ecosystem/grpc-gateway) to handle HTTP <->
+gRPC transcoding.
+
+More Reading:
+
+- https://www.redhat.com/en/blog/comparing-openapi-grpc
+- https://cloud.google.com/blog/products/api-management/understanding-grpc-openapi-and-rest-and-when-to-use-them
+
+### Making Results part of the core Pipeline controller
+
+While we expect that this will be a commonly used component for many
+installations, we do not want to make this a requirement for every installation.
+Users should be able to enable / disable this component freely to fit their
+needs.
+
+### Naming overlap with Task results
+
+There is naming overlap with
+[Task outputs](https://github.com/tektoncd/pipeline/blob/master/docs/tasks.md#emitting-results),
+called Task results. There is a risk that users may confuse Tekton Results with
+Task results.
+
+This is a particularly messy history, because the original proposal for the
+Results API came out before
+[Task results were introduced to the codebase](https://github.com/tektoncd/pipeline/pull/1921)
+and the
+[initial mention](https://github.com/tektoncd/pipeline/issues/1273#issuecomment-546494832)
+of the Task results field (named results to not conflict with PipelineResource
+outputs) was suggested the same week as the original Results API design doc.
+Both of these predated the current TEP process.
```

Comment:
🤣 🤭

https://github.com/tektoncd/pipeline/pull/2706

![image](https://user-images.githubusercontent.com/432502/98594367-8e288a80-22a2-11eb-84cf-13d27b7336ca.png)

fwiw _THIS_ particular feature has been called "result" since at least https://github.com/tektoncd/pipeline/issues/454 (jan 2019)